demos/groktunnel/main.go

```diff
@@ -100,22 +69,57 @@ func main() {
 	}
 }
 
+func serve(vmux *vhost.HTTPMuxer, host, port string) {
+	ml, err := vmux.Listen(net.JoinHostPort(host, port))
+	fatal(err)
+	srv := &http.Server{Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		publicHost := strings.TrimSuffix(net.JoinHostPort(newSubdomain()+host, port), ":80")
+		pl, err := vmux.Listen(publicHost)
+		fatal(err)
+		w.Header().Add("X-Public-Host", publicHost)
+		w.Header().Add("Connection", "close")
+		w.WriteHeader(http.StatusOK)
+		conn, _, _ := w.(http.Hijacker).Hijack()
+		sess := session.New(conn)
+		defer sess.Close()
+		log.Printf("%s: start session", publicHost)
+		go func() {
+			for {
+				conn, err := pl.Accept()
+				if err != nil {
+					log.Println(err)
+					return
+				}
+				ch, err := sess.Open(context.Background())
+				if err != nil {
+					log.Println(err)
+					return
+				}
+				go join(ch, conn)
+			}
+		}()
+		sess.Wait()
+		log.Printf("%s: end session", publicHost)
+	})}
+	srv.Serve(ml)
+}
+
 func join(a io.ReadWriteCloser, b io.ReadWriteCloser) {
-	var wg sync.WaitGroup
-	wg.Add(2)
-	go func() {
-		io.Copy(a, b)
-		wg.Done()
-	}()
-	go func() {
-		io.Copy(b, a)
-		wg.Done()
-	}()
-	wg.Wait()
+	go io.Copy(b, a)
+	io.Copy(a, b)
 	a.Close()
 	b.Close()
 }
 
+func newSubdomain() string {
+	letters := []rune("abcdefghijklmnopqrstuvwxyz1234567890")
+	b := make([]rune, 10)
```

Comment:
This is about ~51 bits. For local use it's plenty, but if someone did host this online they might want more entropy to avoid collisions.
I guess as a demo this doesn't need to be particularly hardened for security, but maybe some warnings about what to watch out for if you host it publicly.
E.g. an attacker could hold open a bunch of sessions to wait for someone else to connect, or try connecting to random endpoints until they find a live one. Putting time limits on sessions could limit the chance of someone discovering one while it's still live.

Reply:
will definitely warn in readme. i'm not sure how well http hijacking even works putting it on like heroku or anything with a balancer